### ISSUE TITLE
(🎁) Support `for`/`else` statement in partially defined check 

### ISSUE THREAD

**Issue by KotlinIsland** [python/mypy#14097]
```py
l: list[int]
for _ in l:
    pass
else:
    result = False
print(result)  # error: Name "result" may be undefined  [partially-defined]
```

This is incorrect, `result` will always be defined in this case.

**Comment by ilinum** [comment]
Thanks for the bug report! 

mypy isn't smart enough to detect that the list `l` is always empty. Note that technically the body of the loop is unreachable but it's not flagged with `--warn-unreachable` ([mypy-play](https://mypy-play.net/?mypy=latest&python=3.11&flags=warn-unreachable)). 

Do you have a real-world example where this would matter? In the toy example here it seems like it's better to just simplify everything to `result = False` and remove the loop.

**Comment by KotlinIsland** [comment]
> mypy isn't smart enough to detect that the list l is always empty.

This issue isn't regarding an empty list, the `else` branch will always be executed here regardless of if `l` has items or not, and `result` will always be set to `False`

> Note that technically the body of the loop is unreachable

Again, this isn't regarding a list that is empty, so the body is potentially reachable.

> Do you have a real-world example where this would matter? 

Yes, some projects I have lit up with false positives when I updated mypy. Basically any scenario where a `for`/`else` is used to set a fallback value.

**Comment by ilinum** [comment]
Yep! You're totally right. Thanks for pointing this out; sent a PR with the fix.

**Comment by KotlinIsland** [comment]
follow up: #14209
